### PR TITLE
DD-1438 dd-manage-deposits: Fix problems found DD-1419

### PR DESCRIPTION
**Pull request by aliassheikh**
Fixes DD-1438 dd-manage-deposits: Fix problems found DD-1419

# Description of changes
- **Changed**  'empty string' is replaced by 'undefined' representation in a text/csv query when column data is null
- **Changed** during adding a new DepositProperties record, its field 'creation-date' will be only filled when it is present in deposit.properties file (otherwise it is null)
- **Changed**  Where possible runtime exceptions are replaced by log warning/error
- **Fixed**: Handling DateTimeExceptions when an invalid  string is parsed by DateTime classes
- **Added** some new Junit tests
- **Improvement**: will give an warning when a deposit.properties file is placed directly in one of the base root folders (see config.yml file) 
- **Improvement**: removed unused code
  
# How to test
  - Starting vagrant
    - `start-preprovisioned-box.py -s dev_archaeology`
    - `deploy.py -m dd-manage-deposit dev_archaeology`
  
  - Deposit vaild/invalid datasets using
    - cd to local folder: `~/git/dans-core-systems/modules/dd-dans-sword2-examples` 
    - run example: `./run-simple-deposit.sh https://dev.sword2.archaeology.datastations.nl/collection/1 user001 user001 src/main/resources/example-bags/valid/audiences`
    
  - Testing a dataset with multiple deposit.properties files (only one database record should be created)
    -   Unzip attached  '14547798/c15ee79c-d6e3-46d0-aa63-bf8ea23fd310.zip' to local vagrant 'shared' folder
    - In the Terminal open vagrant ssh (`vagrant ssh dev_archaeology`)
    - Copy the unzipped folder to '/var/opt/dans.knaw.nl/tmp/sword2-uploads'. For example: 
      - `cp -r /vagrant/provisioning/shared//c15ee79c-d6e3-46d0-aa63-bf8ea23fd310/ /var/opt/dans.knaw.nl/tmp/sword2-
    - uploads` 
   
  - Looking at the results:
    - in the terminal:
      -    open vagrant ssh (`vagrant ssh dev_archaeology`)
      - `tree -L 5 /var/opt/dans.knaw.nl/tmp` or tree -L 6 /var/opt/dans.knaw.nl/tmp`
    - Reports:
      - in the terminal: `curl 'accept: text/csv' http://dev.sword2.archaeology.datastations.nl:20355/report`  
      - in the firefox browser: `http://dev.sword2.archaeology.datastations.nl:20355/report`
      
# Related PRs

(Add links)

* [File Alteration Observer - apatch common library](https://commons.apache.org/proper/commons-io/javadocs/api-release/index.html?org/apache/commons/io/monitor/package-summary.html)

# Notify

@DANS-KNAW/core-systems
[c15ee79c-d6e3-46d0-aa63-bf8ea23fd310.zip](https://github.com/DANS-KNAW/dd-manage-deposit/files/14547798/c15ee79c-d6e3-46d0-aa63-bf8ea23fd310.zip)

### CODE REVIEW

File: src/main/java/nl/knaw/dans/managedeposit/core/DepositProperties.java [hunk 1]
```diff
@@ -172,4 +173,20 @@ public boolean equals(Object o) {
     public int hashCode() {
         return 31 * depositId.hashCode() + depositor.hashCode();
     }
+
+    @Override
+    public String toString() {
+        return "DepositProperties{" +
+            "depositor='" + depositor + '\'' +
```
Comment (jo-pol) [inline src/main/java/nl/knaw/dans/managedeposit/core/DepositProperties.java]:
```suggestion
            "depositor='" + depositor + "'" +
```
looks clearer to me than escaping.

File: src/main/java/nl/knaw/dans/managedeposit/core/service/DepositPropertiesAssembler.java [hunk 2]
```diff
@@ -34,37 +35,32 @@ class DepositPropertiesAssembler {
...
             configuration = DepositPropertiesFileReader.readDepositProperties(depositPropertiesFile);
+            String creationTime = configuration.getString("creation.timestamp");
 
             dp = new DepositProperties(depositPath.getFileName().toString(),
                 configuration.getString("depositor.userId", ""),
                 configuration.getString("bag-store.bag-name", ""),
                 configuration.getString("state.label", ""),
                 TextTruncation.stripEnd(configuration.getString("state.description", ""), TextTruncation.MAX_DESCRIPTION_LENGTH),
-                OffsetDateTime.parse(configuration.getString("creation.timestamp", "")),
+                (creationTime == null || creationTime.isEmpty()) ? null : OffsetDateTime.parse(creationTime),
```
Comment (jo-pol) [inline src/main/java/nl/knaw/dans/managedeposit/core/service/DepositPropertiesAssembler.java:52]: Not part of this PR, but can´t the stripBegin below cause unexpected results and what about trim rather than stripEnd above.

File: src/main/java/nl/knaw/dans/managedeposit/core/service/DepositStatusUpdater.java [hunk 3]
```diff
@@ -44,25 +44,39 @@ public void onDepositCreate(File depositPropertiesFile) {
...
-        dpObject.ifPresent(depositPropertiesDAO::merge);
-        log.debug("onDepositChange: deposit.properties has been changed '{}'", depositPropertiesFile.getParentFile().getAbsolutePath());
+        long folder_size = record.map(DepositProperties::getStorageInBytes).orElse(0L);
+        Optional<DepositProperties> dpObject = depositPropertiesAssembler.assembleObject(depositPropertiesFile, folder_size);
+        if (dpObject.isPresent()) {
+            depositPropertiesDAO.merge(dpObject.get());
+            log.debug("onDepositChange: deposit.properties has been changed '{}'", depositPropertiesFile.getParentFile().getAbsolutePath());
+        }
+        else {
+            writeErrorMsg(depositPropertiesFile.getParentFile().getAbsolutePath());
```
Comment (jo-pol) [inline src/main/java/nl/knaw/dans/managedeposit/core/service/DepositStatusUpdater.java:78]: Good catches! But don´t we need some additional description of the error?

File: src/main/java/nl/knaw/dans/managedeposit/core/service/DepositStatusUpdater.java [hunk 4]
```diff
@@ -82,9 +94,13 @@ public void onDepositDelete(File depositPropertiesFile) {
                 }
             }
             catch (IOException e) {
-                throw new RuntimeException(e);
+                writeErrorMsg(e.getMessage());
```
Comment (jo-pol) [inline src/main/java/nl/knaw/dans/managedeposit/core/service/DepositStatusUpdater.java:97]: Again, is the error message clear enough without more context?

File: src/main/java/nl/knaw/dans/managedeposit/db/DepositPropertiesDAO.java [hunk 5]
```diff
@@ -106,44 +110,53 @@ private Predicate buildQueryCriteria(Map<String, List<String>> queryParameters,
...
+                        break;
+
+                    case "startdate":
+                    case "enddate":
+                        if (value.isEmpty()) {
+                            orPredicateItem = criteriaBuilder.isNull(root.get("depositCreationTimestamp"));
+                        }
+                        else {
+                            try {
+                                DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
```
Comment (jo-pol) [inline src/main/java/nl/knaw/dans/managedeposit/db/DepositPropertiesDAO.java:140]: I fail to see logic to pair up start-dates and end-dates for multiple periods nor to prevent multiple periods. Unit tests would be interesting but far from trivial.

File: src/main/java/nl/knaw/dans/managedeposit/resources/DepositPropertiesResource.java [hunk 6]
```diff
@@ -45,9 +45,11 @@ private String writeHelpInfoText() {
                 Usage:\s
                   - Create reports: GET  basePath/report\s
                   - Clean database: POST basePath/delete-deposit\s
-                    Query string parameters: user, state, startdate, enddate\s
-                    'startdate'/'enddate' format: yyyy-MM-dd\s
-                    Possible 'state' value: ARCHIVED, DRAFT, FAILED, FINALIZING, INVALID, REJECTED, SUBMITTED, UPLOADED, PUBLISHED\s
+                  - Query string parameters: user, state, startdate, enddate, deleted\s
+                    - 'startdate'/'enddate' format: yyyy-MM-dd\s
+                    - 'deleted' is a boolean with values: 'true' or 'false'\s
+                    - Possible 'state' values: ARCHIVED, DRAFT, FAILED, FINALIZING, INVALID, REJECTED, SUBMITTED, UPLOADED, PUBLISHED\s
```
Comment (jo-pol) [inline src/main/java/nl/knaw/dans/managedeposit/resources/DepositPropertiesResource.java:51]: I suppose there is some enum. Better use it.

**Review by jo-pol — COMMENTED** [review]
Did not catch problems. Most remarks are not even related to the issue or deserve a new pull request otherwise. 

Just give logging enough context at the very first line for a reader who is not very familiar with the code.

File: src/main/java/nl/knaw/dans/managedeposit/core/service/DepositPropertiesAssembler.java (hunk 2 above)
Comment (aliassheikh) [reply src/main/java/nl/knaw/dans/managedeposit/core/service/DepositPropertiesAssembler.java:52]: 1- about stripBegin: It strips the path only if it is too long, and a path can´t be empty in this case.
2- about trim and stripEnd:
  - Java says: trim returns a string whose value is this string, with all leading and trailing space removed
  - stripEnd cuts the end part of a text (description) when it is longer than the defined size.
  - Maybe your point is only about the name of the function (changing ´stripEnd´ to trim)???

File: src/main/java/nl/knaw/dans/managedeposit/core/service/DepositStatusUpdater.java (hunk 3 above)
Comment (aliassheikh) [reply src/main/java/nl/knaw/dans/managedeposit/core/service/DepositStatusUpdater.java:78]: Good point.
1- inside writeErrorMsg(String), the function adds extra information:
  `log.error("Error creating / Updating deposit record: '{}'", additionalInfo);`
2- I Need to create and handle a new issue/ticket to handle a few remained exceptions in the source code and also to handle the situations when a deposit does not contain a deposit.properties file or when a deposit.properties file is placed in the root/base folder(s) of the service. I will have a review of all these messages together.

File: src/main/java/nl/knaw/dans/managedeposit/core/service/DepositStatusUpdater.java (hunk 4 above)
Comment (aliassheikh) [reply src/main/java/nl/knaw/dans/managedeposit/core/service/DepositStatusUpdater.java:97]: See please, my previous comment.

File: src/main/java/nl/knaw/dans/managedeposit/db/DepositPropertiesDAO.java (hunk 5 above)
Comment (aliassheikh) [reply src/main/java/nl/knaw/dans/managedeposit/db/DepositPropertiesDAO.java:140]: **You are right.** 
It was in the definition of this module.
[dd-manage-deposite API](https://dans-knaw.github.io/dd-manage-deposit/api.html) :
1- be able to have a collection of deposits from: startdate
2- be able to have a collection of deposits before: enddate

**And you are also right**, this function doesn't take additional check(s), but asks directly the database with this/these specified dates by end user.

Are additional steps necessary?

File: src/main/java/nl/knaw/dans/managedeposit/resources/DepositPropertiesResource.java (hunk 6 above)
Comment (aliassheikh) [reply src/main/java/nl/knaw/dans/managedeposit/resources/DepositPropertiesResource.java:51]: I'm not sure about this one.
I have copied these values from the ingest-flow (also in dd-sowrd2) module where they are used to write the state label of the deposits in the their deposit.properties files. 
This module (dd-manage-deposits) doesn't do anything with these values. It copied the deposit state label from the deposit.properties file as a column of the records.
Then end users can make a selection based of one of these state values.
Here, I tried only to help the end user with the values.

File: src/main/java/nl/knaw/dans/managedeposit/core/DepositProperties.java (hunk 1 above)
Comment (aliassheikh) [reply src/main/java/nl/knaw/dans/managedeposit/core/DepositProperties.java]: okay. I'm committing your suggestion. Thanks.

File: src/main/java/nl/knaw/dans/managedeposit/core/service/DepositPropertiesAssembler.java (hunk 2 above)
Comment (jo-pol) [reply src/main/java/nl/knaw/dans/managedeposit/core/service/DepositPropertiesAssembler.java:52]: I must have been confused assuming stripEnd was applied to a path.

File: src/main/java/nl/knaw/dans/managedeposit/resources/DepositPropertiesResource.java (hunk 6 above)
Comment (jo-pol) [reply src/main/java/nl/knaw/dans/managedeposit/resources/DepositPropertiesResource.java:51]: OK, indeed better not to introduce a maintenance risk (when ingest-flow or dd-sowrd2 change) by enforcing possible values.

File: src/main/java/nl/knaw/dans/managedeposit/db/DepositPropertiesDAO.java (hunk 5 above)
Comment (jo-pol) [reply src/main/java/nl/knaw/dans/managedeposit/db/DepositPropertiesDAO.java:140]: To keep it simple return code 400 ("invalid or incorrectly formatted parameters ... " as defined by the API) in case of repeated start-date or end-date

**Review by jo-pol — COMMENTED** [review]
looks good to me.
To prevent forgetting about it: create new issues before you resolve the open conversations

**Review by jo-pol — APPROVED** [review]

File: src/main/java/nl/knaw/dans/managedeposit/resources/DepositPropertiesResource.java (hunk 6 above)
Comment (aliassheikh) [reply src/main/java/nl/knaw/dans/managedeposit/resources/DepositPropertiesResource.java:51]: Check, it is a good idea.
